### PR TITLE
Use data types for all parameters

### PR DESCRIPTION
**Pull request by smortex**
#### Pull Request (PR) description

While some parameters have proper data types, other are missing one.
Ensure all parameters have a data type.

This refactoring helped spotting optional parameters with a default
value set to "nil" and which have been made really optional with a
default value of undef.

#### This Pull Request (PR) fixes the following issues
n/a

### CODE REVIEW

File: manifests/env.pp [hunk 1]
```diff
@@ -6,21 +6,29 @@
...
+  Boolean                   $cron_always_mail = false,
+  Struct[
+    {
+      minute   => Optional[Variant[Integer, String]],
+      hour     => Optional[Variant[Integer, String]],
+      monthday => Optional[String],
+      month    => Optional[String],
+      week     => Optional[String],
+    }
+  ]                        $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
```
Comment (ekohl) [inline manifests/env.pp:31]: Perhaps this should be extracted to a separate type?

File: manifests/env.pp [hunk 2]
```diff
@@ -6,21 +6,29 @@
...
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
+  Enum['present', 'absent'] $ensure           = 'present',
+  Array[String]             $makeopts         = [],
+  Optional[String::Abspath] $makefile         = undef,
+  String                    $version          = '10.0-RELEASE',
+  String                    $arch             = 'amd64',
+  String                    $jail             = $name,
+  Integer                   $paralleljobs     = $facts['processors']['count'],
```
Comment (ekohl) [inline manifests/env.pp:15]: Does this have a lower bound of 1?

**Review by ekohl — COMMENTED** [review]
Not a real review, just some things that jumped out to me.

File: manifests/env.pp (hunk 1 above)
Comment (smortex) [reply manifests/env.pp:31]: Indeed!  I wanted to do it at some point and forgot about it.

File: manifests/env.pp (hunk 2 above)
Comment (smortex) [reply manifests/env.pp:15]: I guess so, :wink:

**Comment by smortex** [comment]
@ekohl, I adjusted the points you mentioned. Thanks!

File: manifests/init.pp [hunk 3]
```diff
@@ -3,38 +3,38 @@
...
-  String $allow_make_jobs            = '',
-  String $url_base                   = '',
-  String $max_execution_time         = '',
-  String $nohang_time                = '',
-  String $http_proxy                 = '',
-  String $ftp_proxy                  = '',
-  String $build_as_non_root          = '',
-  $environments                      = {},
-  $portstrees                        = {},
+  String                       $zpool                 = 'tank',
```
Comment (igalic) [inline manifests/init.pp:6]: shouldn't this default to zroot by now?

File: manifests/init.pp [hunk 4]
```diff
@@ -3,38 +3,38 @@
...
+  Stdlib::Absolutepath         $zrootfs               = '/poudriere',
+  String                       $freebsd_host          = 'http://ftp6.us.freebsd.org/',
```
Comment (igalic) [inline manifests/init.pp:8]: if we remove set this to `"https://downloads.freebsd.org"`
which is automatically selected based on geographical location

File: types/cron_interval.pp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# The type of job
+type Poudriere::Cron_interval = Struct[
```
Comment (igalic) [inline types/cron_interval.pp:2]: I'm surprised this kind of type doesn't exist in any of our dependencies

File: manifests/portstree.pp [hunk 6]
```diff
@@ -3,13 +3,21 @@
...
+  Optional[String[1]]        $branch           = undef,
+  Enum[
+    'git',
+    'null',
+    'portsnap',
+    'svn',
+    'svn+file',
+    'svn+http',
+    'svn+https',
+  'svn+ssh']  $fetch_method     = 'svn',
```
Comment (igalic) [inline manifests/portstree.pp:17]: this enum could be moved into its own type definition

**Review by igalic — COMMENTED** [review]
🙋🏻‍♀️

File: manifests/init.pp (hunk 3 above)
Comment (smortex) [reply manifests/init.pp:6]: I had to set it to `zroot` on the system I am configuring.  My other poudriere instance use `zhome`.  I trend to think that there should be no default value…  There are a bunch of other places where the default value / behavior was surprising for me, but I refrained to change this in this PR.

I guess I'll open an issue to discuss this in more details with the maintainers and build a plan to improve this :wink:

File: manifests/init.pp (hunk 4 above)
Comment (smortex) [reply manifests/init.pp:8]: Totally agree, I override this in my control-repo…  See my previous comment.

File: manifests/portstree.pp (hunk 6 above)
Comment (smortex) [reply manifests/portstree.pp:17]: It's used a single time to I was not happy with it, but the horror required by the buggy puppet-strings to cope with this make me think we definitively SHALL do it…

Comment (smortex) [reply manifests/portstree.pp]: Fixed!

File: manifests/env.pp [hunk 7]
```diff
@@ -6,21 +6,21 @@
...
-  $pkg_makeopts     = {},
-  $pkg_optsdir      = nil,
-  $portstree        = 'default',
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
+  Enum['present', 'absent'] $ensure           = 'present',
+  Array[String]             $makeopts         = [],
+  Optional[String::Abspath] $makefile         = undef,
+  String                    $version          = '10.0-RELEASE',
```
Comment (igalic) [inline manifests/env.pp:12]: nope… this should default to the Host OS version
or is that also something you'll fix in a new round?

**Review by igalic — APPROVED** [review]
this is looking somewhat better

File: manifests/env.pp (hunk 7 above)
Comment (bastelfreak) [reply manifests/env.pp:12]: if it should be changed, it should happen in another  PR, as this one only focuses on data types.

File: manifests/env.pp [hunk 8]
```diff
@@ -6,21 +6,21 @@
...
-  $pkgs             = [],
-  $pkg_file         = nil,
-  $pkg_makeopts     = {},
-  $pkg_optsdir      = nil,
-  $portstree        = 'default',
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
+  Enum['present', 'absent'] $ensure           = 'present',
+  Array[String]             $makeopts         = [],
```
Comment (bastelfreak) [inline manifests/env.pp:10]: I guess empty strings aren't allowed?
```suggestion
  Array[String[1]]          $makeopts         = [],
```

File: manifests/init.pp [hunk 9]
```diff
@@ -3,38 +3,38 @@
...
+  String                       $no_restricted         = '',
+  String                       $allow_make_jobs       = '',
+  String                       $url_base              = '',
+  String                       $max_execution_time    = '',
+  String                       $nohang_time           = '',
+  String                       $http_proxy            = '',
+  String                       $ftp_proxy             = '',
+  String                       $build_as_non_root     = '',
+  Hash                         $environments          = {},
+  Hash                         $portstrees            = {},
```
Comment (bastelfreak) [inline manifests/init.pp:37]: I'm not sure if I like the whole indentation. It adds a lot of noise to the PR and future changes to the datatypes or new parameters might require more whitespace on all the other params.

File: manifests/env.pp (hunk 8 above)
Comment (smortex) [reply manifests/env.pp:10]: Yes, a bunch of `String` should be `String[1]`

File: manifests/init.pp (hunk 9 above)
Comment (smortex) [reply manifests/init.pp:37]: I had this in 2 commits, one for the actual changes and another to align fields properly.
I fsck'ed up this when integrating @igalic suggestions, so I will re-do this cleanly with your suggestions.

File: manifests/env.pp (hunk 7 above)
Comment (smortex) [reply manifests/env.pp]: I opened a PR with a list of the weird things I stumbled on and can help fixing.  This item is in the list!

**Review by igalic — APPROVED** [review]
this looks better

File: types/cron_interval.pp [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# The type of job
```
Comment (bastelfreak) [inline types/cron_interval.pp:1]: can you add puppet-strings documentation for the two new types?
```
# @summary ...
```

should be sufficient (same for the other new type

Comment (smortex) [reply types/cron_interval.pp]: Outch!

Comment (smortex) [reply types/cron_interval.pp]: (for the comment I mean… copy/paste/forget-to-remove-or-adjust)

**Review by bastelfreak — APPROVED** [review]
looks good to me, but @igalic should probably take a look as well.

File: manifests/env.pp [hunk 11]
```diff
@@ -6,21 +6,21 @@
...
-  $pkg_optsdir      = nil,
-  $portstree        = 'default',
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
+  Enum['present', 'absent'] $ensure           = 'present',
+  Array[String[1]]          $makeopts         = [],
+  Optional[String::Abspath] $makefile         = undef,
+  String[1]                 $version          = '10.0-RELEASE',
+  String[1]                 $arch             = 'amd64',
```
Comment (igalic) [inline manifests/env.pp:13]: this is a known set that could also be an enum, but we can move this down to the better defaults refactoring

File: types/fetch_method.pp [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# @summary A supported fetch method for poudriere
+type Poudriere::Fetch_method = Enum[
+  'git',
+  'null',
+  'portsnap',
+  'svn',
+  'svn+file',
+  'svn+http',
+  'svn+https',
+  'svn+ssh',
```
Comment (igalic) [inline types/fetch_method.pp:10]: there's also `src=$AbsolutePath` but I'm not sure how to model that in puppet

File: manifests/env.pp (hunk 11 above)
Comment (smortex) [reply manifests/env.pp:13]: That list changed a few times in the last years… I think `sparc64` is not an option anymore, `arm` and `arm64` where added more recently.  I am not sure if the `Enum` is the best in this case, but I do not have strong opinions about it.

File: types/fetch_method.pp (hunk 12 above)
Comment (smortex) [reply types/fetch_method.pp:10]: I don't see this method in `poudriere-ports(8)` which `$fetch_method` correspond to.  However, it's an option for `poudriere-jail(8)` (along with `tar=...`, `url=...`), are you mixing these two `-m` arguments or am I reading an outdated man page?

Comment (igalic) [reply types/fetch_method.pp:10]: no… I'm using poudriere from git HEAD 😅 to only build pkgbase 

let's leave this as it is right now

File: manifests/env.pp (hunk 11 above)
Comment (igalic) [reply manifests/env.pp:13]: and this module might not be the best place for it, maybe something closer to the FreeBSD system could carry that info / type

**Comment by smortex** [comment]
Hey, can we merge this?  Maybe @xaque208 wants to review?

File: manifests/env.pp [hunk 13]
```diff
@@ -6,21 +6,21 @@
...
-  $pkg_file         = nil,
-  $pkg_makeopts     = {},
-  $pkg_optsdir      = nil,
-  $portstree        = 'default',
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 0, monthday => '*', month => '*', weekday => '*' },
+  Enum['present', 'absent'] $ensure           = 'present',
+  Array[String[1]]          $makeopts         = [],
+  Optional[String::Abspath] $makefile         = undef,
```
Comment (kenyon) [inline manifests/env.pp:11]:
```suggestion
  Optional[Stdlib::Absolutepath] $makefile         = undef,
```

https://github.com/puppetlabs/puppetlabs-stdlib#stdlibabsolutepath

File: manifests/env.pp [hunk 14]
```diff
@@ -6,21 +6,21 @@
...
+  String[1]                 $version          = '10.0-RELEASE',
+  String[1]                 $arch             = 'amd64',
+  String[1]                 $jail             = $name,
+  Integer[1]                $paralleljobs     = $facts['processors']['count'],
+  Array[String[1]]          $pkgs             = [],
+  Optional[Stdlib::Abspath] $pkg_file         = undef,
```
Comment (kenyon) [inline manifests/env.pp:17]:
```suggestion
  Optional[Stdlib::Absolutepath] $pkg_file         = undef,
```

File: manifests/env.pp [hunk 15]
```diff
@@ -6,21 +6,21 @@
...
+  Hash                      $pkg_makeopts     = {},
+  Optional[Stdlib::Abspath] $pkg_optsdir      = undef,
```
Comment (kenyon) [inline manifests/env.pp:19]:
```suggestion
  Optional[Stdlib::Absolutepath] $pkg_optsdir      = undef,
```

File: manifests/portstree.pp [hunk 16]
```diff
@@ -3,13 +3,13 @@
...
 
 define poudriere::portstree (
-  $ensure           = 'present',
-  $portstree        = $name,
-  $branch           = undef,
-  $fetch_method     = 'svn',
-  $cron_enable      = false,
-  $cron_always_mail = false,
-  $cron_interval    = { minute => 0, hour => 22, monthday => '*', month => '*', week => '*' },
+  Enump['present', 'absent'] $ensure           = 'present',
```
Comment (kenyon) [inline manifests/portstree.pp:6]:
```suggestion
  Enum['present', 'absent'] $ensure           = 'present',
```

**Review by kenyon — CHANGES_REQUESTED** [review]

**Comment by kenyon** [comment]
Test coverage could certainly be better, if typos in data types aren't causing test failures. :grimacing:

**Review by kenyon — APPROVED** [review]

**Comment by smortex** [comment]
> Test coverage could certainly be better, if typos in data types aren't causing test failures. :grimacing:

Yes, good catch!  I do not use these classes on my home setup and since the test suite was green, I considered it was fine…

**Comment by smortex** [comment]
I extended the test suite and am discovered more errors.  I am fixing them ATM!

**Review by xaque208 — APPROVED** [review]
Nice work, this looks good to me.  Thanks for putting this together @smortex.  What is the syntax `String[1]` doing?

**Comment by smortex** [comment]
@xaque208 `String[1]` ensure that the string is at least one char long… so an empty string is not allowed :wink:

File: types/fetch_method.pp [hunk 17]
```diff
@@ -0,0 +1,11 @@
+# @summary A supported fetch method for poudriere
+type Poudriere::Fetch_method = Enum[
+  'git',
```
Comment (igalic) [inline types/fetch_method.pp:3]: git can have almost the same +es as svn:

```
git, # sadly the default with this is the git protocol
git+http, # bad
git+https,
git+ssh,
```

i also have this pr https://github.com/freebsd/poudriere/pull/813

**Review by igalic — CHANGES_REQUESTED** [review]
just one more thing i noticed

File: types/fetch_method.pp (hunk 17 above)
Comment (smortex) [reply types/fetch_method.pp:3]: Yes, git is becommind a thing I guess with the switch to the git VCS.  I adjusted the list accordingly, thanks!

**Review by igalic — APPROVED** [review]

**Pull request merged**